src/App.tsx

```diff
@@ -1,24 +1,234 @@
+
+  const showError = (errorType: ErrorMessage) => {
+    setIsError(true);
+    setError(errorType);
+    setTimeout(() => {
+      setIsError(false);
+      setError(null);
+    }, 3000);
+  };
+
+  const filteredTodos = useMemo(() => {
+    switch (filterBy) {
+      case Filter.active:
+        return todos.filter(todo => !todo.completed);
+      case Filter.completed:
+        return todos.filter(todo => todo.completed);
+      default:
+        return todos;
+    }
+  }, [filterBy, todos]);
+
+  const isAllCompleted
+    = todos.filter(todo => todo.completed).length === todos.length;
+
+  const isAllUncompleted
+    = todos.filter(todo => !todo.completed).length === todos.length;
+
+  const resetField = (): void => {
+    setTodoTitle('');
+    setIsError(false);
+    setError(null);
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setProcessingIds(current => [...current, todoId]);
+    deleteTodo(todoId)
+      .then(() => setTodos(
+        currentTodos => currentTodos.filter(todo => todo.id !== todoId),
+      ))
+      .catch(() => {
+        showError(ErrorMessage.deleteError);
+      })
+      .finally(() => setProcessingIds([]));
+  };
+
+  const handleUpdateTodo = (todo: Todo) => {
+    setProcessingIds(current => [...current, todo.id]);
+    const currentTodo = todos.filter(toDo => toDo.id === todo.id)[0];
+    const indexOfTodo = todos.indexOf(currentTodo);
+
+    updateTodo(todo)
+      .then(() => setTodos(
+        currentTodos => [
+          ...currentTodos.slice(0, indexOfTodo),
+          todo,
+          ...currentTodos.slice(indexOfTodo + 1),
+        ],
+      ))
+      .catch(() => {
+        showError(ErrorMessage.updateError);
+      })
+      .finally(() => setProcessingIds([]));
+  };
+
+  const clearCompleted = () => {
+    todos.map(todo => {
+      if (todo.completed) {
+        setProcessingIds(current => [...current, todo.id]);
+        handleDeleteTodo(todo.id);
+      }
+
+      return todo;
+    });
+  };
+
+  const handleToggleAll = () => {
+    if (isAllCompleted || isAllUncompleted) {
+      todos.map(todo => handleUpdateTodo({
+        ...todo,
+        completed: !todo.completed,
+      }));
+    } else {
+      todos.map(todo => !todo.completed && handleUpdateTodo({
+        ...todo,
+        completed: !todo.completed,
+      }));
+    }
+  };
+
+  const handleAddTodo = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!todoTitle.trim()) {
+      showError(ErrorMessage.emptyTitle);
+
+      return;
+    }
+
+    const newTodo: Todo = {
+      id: Math.max(...todos.map(todo => todo.id)) + 1,
+      title: todoTitle,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({ ...newTodo, id: 0 });
+
+    setProcessingIds(current => [...current, 0]);
+    createTodo(newTodo)
+      .then(newToDo => {
+        setTodos(currentTodos => [...currentTodos, newToDo]);
+        resetField();
+      })
+      .catch(() => {
+        setTempTodo(null);
```

Comment:
redundant, it's in finally

Reply:
not fixed
<img width="470" alt="image" src="https://github.com/mate-academy/react_todo-app-with-api/assets/98764641/a988f652-d5ba-4309-9386-3ad15b307cf8">